packages/runtime/container-runtime/src/blobManager.ts

```diff
@@ -106,87 +163,196 @@ export class BlobManager {
         return `/${BlobManager.basePath}/${blobId}`;
     }
 
-    public async getBlob(blobId: string): Promise<IFluidHandle<ArrayBufferLike>> {
-        const storageId = this.redirectTable?.get(blobId) ?? blobId;
-        assert(this.hasBlob(storageId), 0x11f /* "requesting unknown blobs" */);
+    private get storageIds(): Set<string> {
+        const ids = new Set<string | undefined>(this.redirectTable.values());
+
+        // If we are detached, we will not have storage IDs, only undefined
+        const havePendingBlobs = ids.delete(undefined);
+
+        // We should not have undefined entries in the table unless we are detached
+        assert(!havePendingBlobs || this.runtime.attachState === AttachState.Detached && ids.size === 0,
+            "redirect table must not have an undefined value while attached");
```

Comment:
The assert is stronger than what is stated: it asserts that the redirect table for a detached container either has storage ids assigned for all entries (passes !havePendingBlobs) or storage ids assigned for zero entries (passes size === 0).

That could use some additional explanation.  Is this a special case for the "walk up and use" scenario that doesn't apply to a disconnected (but previously attached) container?

Reply:
The redirect table should always have only undefined values while detached and only defined values while attached, since when attached I am not putting the IDs into the table until the BlobAttach op round-trips. Detached is different due to IDetachedBlobStorage, so I am just putting these IDs as keys with undefined values in the table since BlobManager doesn't track the upload of these blobs.

Reply:
Thanks.  I did my best to update the comment & assert, but you'll want to review.